text/2-appetizers/3-errors.md

```diff
@@ -44,6 +44,25 @@ fn add(origin, val_to_add: u32) -> DispatchResult {
 
 Notice that the `Error` type always takes the generic parameter `T`. Notice also that we have verified all preconditions, and thrown all possible errors before ever writing to storage.
 
+## Throwing Generic Errors
+
+`DispatchResult` returned from the above dispatchable call is defined as [`Result<(), DispatchError>`](https://substrate.dev/rustdocs/master/frame_support/dispatch/type.DispatchResult.html). Looking up on [`DispatchError`](https://substrate.dev/rustdocs/master/frame_support/dispatch/enum.DispatchError.html), you will see it is an enum containing a variant of `Other(&'static err)`.
+
+This means you can also throw generic error created from mere string (your error message) as follows:
+
+```rust, ignore
+fn add(origin, val_to_add: u32) -> DispatchResult {
+	// -- snip --
+	let result = match Self::sum().checked_add(val_to_add).ok_or("addition overflow")?;
+	Sum::put(result);
+	Ok(())
+}
+```
+
+Notice the pattern of `.ok_or("my error message")?;`. This is really handy when you have a function call that returns an `Option` and you expect there should be a value inside. If not, returns early with an error message, all the while unwrapping the value for your further processing.
+
+If your function returns a `Result<T, E>`, you could apply `.map_err(|_e| "my error message")?;` in the same spirit.
+
```

Comment:
This information is correct and well-written. However, I don't think it should be included here in the Appetizers, a section that teaches the fundamentals to beginners. I have two reasons for this.
1. Beginners who read this will likely not know whether they should use `decl_error!` or `DispatchResult::Custom`. That causes uncertainty and may cause them to feel overwhelmed and give up.
2. I'm not sure `DispatchResult::Custom` is ever a better choice than `decl_error!`. Maybe some more advanced substrate hackers know more about that (@bkchr , @gnunicorn , @tomusdrw ).

I think the options for where to include this info about `DispatchError::Custom` are:
1. Nowhere
2. In a separate entree recipe specifically about runtime errors

Reply:
Agree with what you said. This is something we can look at from both sides.

1. Not every kind of error I want to declare a custom error for. Sometimes I just want to make sure the computation is okay and if not, return early with error. Basically every time when we want to use `unwrap`, we can always instead use one of these two calls.

2. When dealing with Result and Option and check for the result, except for those that lead to later significantly different handling, checking the result using match statement to me is like interrupting my overall logic flow with error handling. The ok_or and map_err smooth this out.

But I am fine either putting this section here, or somewhere else.

Reply:
I agree with point 2 completely. `ok_or` and `map_err` are great. They do work with  custom errors though. [ok_or example](https://github.com/paritytech/substrate/blob/master/frame/balances/src/lib.rs#L1026) [map_err example](https://github.com/paritytech/substrate/blob/master/frame/treasury/src/lib.rs#L333-L334)

How about this idea: In this Appetizer on errors, we _do_ introduce `ok_or` and `map_err` similarly to how you've done here, but _not_ `DispatchError::Other`. I'm working t oget the recipes to a state so that every piece of code shown in the text is available in-context in an actual buildable pallet, so that means we'll need to add the new syntax to the adding machine pallet. A function like this should do:
```rust
fn add_alternate(origin, val_to_add: u32) -> DispatchResult {
	let _ = ensure_signed(origin)?;

	// First check for unlucky number 13
	ensure!(val_to_add != 13, <Error<T>>::UnluckyThirteen);

	// Now check for overflow while adding
	let result = Self::sum().checked_add(val_to_add).ok_or(<Error<T>>::SumTooLarge))?;

	// Write the new sum to storage
	Sum::put(result);

	Ok(())
}
```

Then it's up to you whether you whether you want to introduce `DispatchError::Other` in another recipe. I'm fine either way on that.

Reply:
Updated.